plugins/inputs/jenkins/jenkins.go

```diff
@@ -0,0 +1,440 @@
+## this config will limit the layers of pull, default value 0 means
+## unlimited pulling until no more sub jobs
+#  max_subjob_depth = 0
+## in workflow-multibranch-plugin, each branch will be created as a sub job
+## this config will limit to call only the lasted branches
+## sub jobs fetch in each layer
+#  empty will use default value 10
+#  max_subjob_per_layer = 10
+#  job_exclude = [ "job1", "job2/subjob1/subjob2", "job3/*"]
+
+## Node filter
+#  node_exclude = [ "node1", "node2" ]
+
+## Woker pool for jenkins plugin only
+#  empty this field will use default value 30
+#  max_connections = 30
+`
+
+// measurement
+const (
+	measurementNode = "jenkins_node"
+	measurementJob  = "jenkins_job"
+)
+
+// SampleConfig implements telegraf.Input interface
+func (j *Jenkins) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description implements telegraf.Input interface
+func (j *Jenkins) Description() string {
+	return "Read jobs and cluster metrics from Jenkins instances"
+}
+
+// Gather implements telegraf.Input interface
+func (j *Jenkins) Gather(acc telegraf.Accumulator) error {
+	if j.instance == nil {
+		client, te := j.initClient()
+		if te != nil {
+			return te
+		}
+		if te = j.newInstance(client); te != nil {
+			return te
+		}
+	}
+
+	j.gatherNodesData(acc)
+	j.gatherJobs(acc)
+
+	return nil
+}
+
+func (j *Jenkins) initClient() (*http.Client, error) {
+	tlsCfg, err := j.ClientConfig.TLSConfig()
+	if err != nil {
+		return nil, fmt.Errorf("error parse jenkins config[%s]: %v", j.URL, err)
+	}
+	return &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsCfg,
+			MaxIdleConns:    j.MaxConnections,
+		},
+		Timeout: j.ResponseTimeout.Duration,
+	}, nil
+}
+
+// seperate the client as dependency to use httptest Client for mocking
+func (j *Jenkins) newInstance(client *http.Client) error {
+	// create instance
+	var err error
+	j.instance, err = gojenkins.CreateJenkins(client, j.URL, j.Username, j.Password).Init()
+	if err != nil {
+		return fmt.Errorf("error connect jenkins instance[%s]: %v", j.URL, err)
+	}
+
+	// init job filter
+	j.jobFilter, err = filter.Compile(j.JobExclude)
+	if err != nil {
+		return fmt.Errorf("error compile job filters[%s]: %v", j.URL, err)
+	}
+
+	// init node filter
+	j.nodeFilter, err = filter.Compile(j.NodeExclude)
+	if err != nil {
+		return fmt.Errorf("error compile node filters[%s]: %v", j.URL, err)
+	}
+
+	// init tcp pool with default value
+	if j.MaxConnections <= 0 {
+		j.MaxConnections = 30
+	}
+
+	// default sub jobs can be acquired
+	if j.MaxSubJobPerLayer <= 0 {
+		j.MaxSubJobPerLayer = 10
+	}
+
+	j.semaphore = make(chan struct{}, j.MaxConnections)
+
+	return nil
+}
+
+func (j *Jenkins) gatherNodeData(node *gojenkins.Node, url string, acc telegraf.Accumulator) error {
+
+	badFormatErr := func(url string, field interface{}, want string, fieldName string) error {
+		return fmt.Errorf("error bad format[%s]: fieldName: %s, want %s, got %T", url, fieldName, want, field)
```

Comment:
Too much info here.  If a value is not as expected just return an error like: `fmt.Errorf("error parsing node %s", node)`

Reply:
👍